src/Codeception/Util/ReflectionPropertyAccessor.php

```diff
@@ -105,4 +99,43 @@ public function createWithProperties($class, array $data)
         } while ($class);
         return $obj;
     }
+
+    /**
+     * @param mixed $obj
+     * @param string $class
+     *
+     * @return bool
+     */
+    private function isNoObjectOf($obj, $class)
+    {
+        return false === \is_a($obj, $class);
+    }
+
+    /**
+     * @param array            $data
+     * @param \ReflectionClass $reflectedEntity
+     *
+     * @return mixed
+     * @throws \ReflectionException
+     */
+    private function createObjectViaReflection(ReflectionClass $reflectedEntity, array $data)
+    {
+        $constructorParameters = [];
+        $constructor           = $reflectedEntity->getConstructor();
+        if (null !== $constructor) {
+            foreach ($constructor->getParameters() as $parameter) {
+                if ($parameter->isOptional()) {
+                    $constructorParameters[] = $parameter->getDefaultValue();
+                } elseif (\array_key_exists($parameter->getName(), $data)) {
+                    $constructorParameters[] = $data[$parameter->getName()];
```

Comment:
With this order of `if` branches, does it mean we can't pass a value to optional constructor parameter? It looks like switching `if` and `elseif` would allow that.

Reply:
I've checked it (and will push another test in a minute). Here, we just collect constructor parameters and instantiate the obj with the default value(s). Since the instance continues to run through `setPropertiesForClass` any given value for an optional constructor parameter would be set afterwards.

Reply:
> Since the instance continues to run through setPropertiesForClass any given value for an optional constructor parameter would be set afterwards.

Is my understanding correct that this assumes the constructor does nothing with arguments and just sets private property value? In this case we could just drop whole thing, use `ReflectionClass::newInstanceWithoutConstructor` and then set all properties directly. I think whole point of that code (added by @burned42, IIRC) was to allow full use of constructor. Maybe it would be better to direct this question to him, or just fix order of branches.

Reply:
The reason why I decided to put the 'default parameter' branch first was that I thought this way existing tests, involving classes with only optional parameters, would still work the same way as before (before, no parameters where passed to the constructor). So I was basically just trying to avoid breaking existing tests. If you don't think that's a problem then I guess switching the order of branches would be a good way to go :)